### ISSUE TITLE
Unexpected INFEASIBLE revisited

### ISSUE THREAD

**Issue by saddy001** [google/or-tools#2564]
* Linux, Python, CP-SAT

I have a model that returns INFEASABLE in ortools v9.0 where it gives solutions in v8.2. Are there any model definitions / negligences that were supported in the previous version that are not supported anymore like in #1634?

**Comment by lperron** [comment]
Without the model, I cannot say.

**Comment by saddy001** [comment]
I guess I found the reason. Assume we have
`model.Add(a == b == 0).OnlyEnforceIf(a)`

Where
* a is a BoolVar
* b is an IntVar(0..20)

Now I have the additional condition
`model.Add(a == b != 0).OnlyEnforceIf(a.Not())`

That worked in the previous version. But not in 9.0. Here's what I think it should do:
```
a   b   R1      R2
===================
T   0   T       T/F
T   1   F       T/F
T   2   F       T/F
F   0   T/F     F
F   1   T/F     T
F   2   T/F     T
```
Where R1/R2 are the results of a (True or False), that will be enforced in the first/the second OnlyEnforceIf. 
Am I wrong about how this should work?

**Comment by lperron** [comment]
You cannot write this equation.
It is not parsed correctly.

**Comment by stradivari96** [comment]
A little bit off-topic but what are these checks for?
https://github.com/google/or-tools/blob/0514d810224a89b76b1d3d2d7864e21aa214ce26/ortools/sat/python/cp_model.py#L614-L636 

Without them this does raise the non-implemented error

**Comment by Mizux** [comment]
we want to forbid the use of `if expr:` since it is parsed by the python interpreter prior to the solve so we overload  the `__bool__()` operator.
BUT for certain range/value actually you can perform the conversion because result will be always be True or False so in this case you return the corresponding value.

e.g. if you have a positive expression in bound [0, 0] with a non zero constant value or non zero coefficient then it must be false otherwise it is True (i.e. `0 * x + 0 * y + ... + 0` is in `[0, 0]` ).

**Comment by stradivari96** [comment]
Interesting, so this is what happens:

```
a == b == 0
```
is parsed as
```
(a == b) and (b == 0)
```
as `bool(a == b)` is False, the expression is just evaluated to
```
a == b
```

TLDR:
@saddy001 just don't do `==` 3 times, also, this does not make much sense:
```python
# a => not(a) ??
model.Add(a == b).OnlyEnforceIf(a)
model.Add(b == 0).OnlyEnforceIf(a)
```

**Comment by saddy001** [comment]
> I guess you just want
> 
> ```python
> model.Add(b == 0).OnlyEnforceIf(a)
> model.Add(b != 0).OnlyEnforceIf(a.Not())
> ```
Yes, thanks @lperron and @stradivari96. I had another look in the docs and there is a nice [example](https://developers.google.com/optimization/cp/channeling) explaining my case.
One last question: The second contraint seems redundant to me:
`model.Add(b != 0).OnlyEnforceIf(a.Not())`

Is there any harm in omitting it?

**Comment by lperron** [comment]
enforcing is just an implication, not an equivalence.

it is just a.not() => b != 0

I believe you need both.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
